Add support for Packet

d-mo opened the pull request.

piontec committed:
correctly handle lack of capacity of AWS spot ASGs

d-mo committed:
Add support for Packet (wip)

jbartosik committed:
Update VPA README with description of limit control

d-mo committed:
Drop packngo dependency, use template for cloudinit script

invidian commented on cluster-autoscaler/cloudprovider/packet/README.md:
```diff
@@ -0,0 +1,53 @@
+## Permissions and credentials
+
+The autoscaler needs a `ServiceAccount` with permissions for Kubernetes and
+requires credentials for interacting with OpenStack.
```
OpenStack?

invidian commented on cluster-autoscaler/cloudprovider/packet/README.md:
```diff
@@ -0,0 +1,53 @@
+| cluster-autoscaler-cloud-config | Global/plan        | The Packet plan (aka size/flavor) for new nodes in the nodepool (default: t1.small)                                                |
+| cluster-autoscaler-cloud-config | Global/billing     | The billing interval for new nodes (default: hourly)                                                                               |
+| cluster-autoscaler-cloud-config | Global/os          | The OS image to use for new nodes (default: ubuntu_18_04). If you change this also update cloudinit.                               |
+| cluster-autoscaler-cloud-config | Global/cloudinit   | The base64 encoded init script that installs Docker and kubeadm on newly provisioned nodes. Default value tested with Ubuntu 18.04 |
```
Hm, rather than specifying the script here, could we use cloud-init config here? Or is it both supported, as in https://support.packet.com/kb/articles/user-data ? If so, then description here could be different, since in our case we will use cloud-init format for provisioning new nodes.

invidian commented on cluster-autoscaler/cloudprovider/packet/examples/cluster-autoscaler-secret.yaml:
```diff
@@ -0,0 +1,49 @@
+  usage-bootstrap-signing: "true"
+
+  # Extra groups to authenticate the token as. Must start with "system:bootstrappers:"
+  auth-extra-groups: system:bootstrappers:kubeadm:default-node-token,system:bootstrappers:worker,system:bootstrappers:ingress
```
Missing newline.

d-mo replied:
Fixed

invidian commented on cluster-autoscaler/cloudprovider/packet/packet_cloud_provider.go:
```diff
@@ -0,0 +1,198 @@
+func BuildPacket(opts config.AutoscalingOptions, do cloudprovider.NodeGroupDiscoveryOptions, rl *cloudprovider.ResourceLimiter) cloudprovider.CloudProvider {
+	var config io.ReadCloser
+
+	// Should be loaded with --cloud-config /etc/kubernetes/kube_openstack_config from master node
```
I'm not sure if references to openstack here are correct.

invidian commented on cluster-autoscaler/cloudprovider/packet/packet_manager.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// createPacketManager creates the desired implementation of packetManager.
+// Currently reads the environment variable MAGNUM_MANAGER to find which to create,
```
`s/PACKET_MANAGER/MAGNUM_MANAGER/` here

invidian commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,324 @@
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+		// klog.Fatalf("Error listing nodes: %v", err)
```
Perhaps this line should be removed.

invidian commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,324 @@
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+		// klog.Fatalf("Error creating node: %v", err)
```
Ditto

invidian reviewed (changes requested):
LGTM in general, just found some small nits.

vivekbagade committed:
Adding config for event filtering

CA Scalability improvement by throttling number of events
fired in the case where there are a lot of unschedulable
pods.

d-mo committed:
Add support for hardware reservations

Force user to provide custom token-id & token-secret

invidian commented:
@d-mo can you rebase `master` branch outside of the PR? It will make it easier for reviewing :)

d-mo commented:
@invidian Done

feiskyer committed:
Add cache for vmss VM

feiskyer committed:
Fix Azure client requests stuck issues on http.StatusTooManyRequests

yan234280533 committed:
nodeGroup judy IsNil to avoid crashed

feiskyer committed:
Allow scaling multiple vmss synchronously

feiskyer committed:
Ensure upcoming nodes are different

cclauss committed:
Use print() function in both Python 2 and Python 3

Legacy __print__ statements are syntax errors in Python 3 but __print()__ function works as expected in both Python 2 and Python 3.

k8s-ci-robot committed:
Merge pull request #2146 from vivekbagade/eventagg

Adding config for event filtering

silenceper committed:
change fmt.Errorf to klog.Errorf

k8s-ci-robot committed:
Merge pull request #2152 from feiskyer/bulk-scale-up

Allow scaling multiple Azure vmss synchronously

rata committed:
Use Metrics Server instead of Heapster in the FAQ

Heapster is deprecated and retired now:

https://github.com/kubernetes-retired/heapster/

The Metrics Server should be used now, so let's mention that on the FAQ
to avoid confusions.

aleksandra-malinowska committed:
Make scale down unit tests faster

aleksandra-malinowska committed:
Add resource limit type to NotTriggerScaleUp event

aleksandra-malinowska committed:
Separate running scale up test from checking results

k8s-ci-robot committed:
Merge pull request #2155 from silenceper/master

change fmt.Errorf to klog.Errorf

aleksandra-malinowska committed:
Add no scale up test

k8s-ci-robot committed:
Merge pull request #2151 from feiskyer/fix-stuck

Fix Azure client requests stuck issues on http.StatusTooManyRequests

k8s-ci-robot committed:
Merge pull request #2123 from jbartosik/patch-1

Update VPA README with description of limit control

k8s-ci-robot committed:
Merge pull request #2160 from aleksandra-malinowska/scale-up-events-fix

Add resource limit type to NotTriggerScaleUp event

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+		reservation = "next-available"
+	}
+
+	cr := DeviceCreateRequest{
```
This only supports one pool (as the README says). The autoscaler generic code, however, supports more than one pool and other providers are using more than one pool.

Would it be possible to support multiple pools, maybe using the nodes parameter specified here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca ?

From the top of my head (not autoscaler expert, this is my first time looking at autoscaler code) maybe something like:
```
--nodes 1:10:baremetal_0:could-config-pool-1
--nodes 1:10:c2.medium.x86:could-config-pool-2
--nodes 1:10:baremetal_0:could-config-pool-3
```
So, the format is something like
```
--nodes <min>:<max>:<instance_type>:<cloudConfig key in the secret for this pool> 
```

And support multiple node pools this way (the code should be adapted to work with multiple pools too, of course :))

Also, probably this can even be simpler. I think there is no need to have a cloud-config per pool, so we can just have one in the secret, like we have now, and remove the param. And have something even simpler, like:

```
--nodes 1:10:baremetal_0
--nodes 1:10:c2.medium.x86
--nodes 1:10:other.instance.type
```
And all pools use the same cloud-config (specified as now, in the secret).

If something else is missing, we can add it to either there `--nodes` params or to pre-established keys in a kubernetes secret/configmap. Probably a parameter to specify the poolname is missing in my proposal, for example :).

I think some of the other parameters currently in the kubernetes secret might not be important on a per-pool basis (like facility, as all nodes should be in the same facility and there is no concept of AZ in Packet, but I might be missing something). Some others, like reservations or billing, might be and can be added either to the secret or params.

What do you think? Does it make sense?

d-mo replied:
Hi Rodrigo. Your suggestion makes sense to me. However, I think it was in our call the other day when somebody (perhaps @deitch?) mentioned that it's already possible to support multiple nodepools by running multiple instances of cluster-autoscaler. This makes sense to me as well and requires no extra dev ;)

As a developer I'd rather keep complexity low. On the other hand, as a cluster-autoscaler user, I'd much rather configure it once for multiple nodepools. So, I tend to agree that it's probably worth the extra dev effort and complexity. That said, it's not a trivial change and it doesn't feel like a particularly urgent requirement, so I'd probably schedule this enhancement for the next iteration.

deitch replied:
It was not I on the call, @d-mo , but you can do that. It feels a little "weak" compared to what you do on hyperscalers. E.g. when I run things in AWS (yes, I sometimes do stuff not on Packet, but only when someone insists! :-) ), I set the ASG discovery, it can pick up tags on multiple ASG, and manage each as a separate nodepool.

We _should_ do the same here, i.e. have it have multiple nodepools. 

I do agree that it is a v2 (or v0.2, etc.) request. Let's get it in working with just one nodepool, and immediately open an issue on it requesting multiple nodepools.

rata replied:
Thanks a lot for the answer, to both! :)

Deploying the autoscaler for each pool (we have installations with 10 or more pools :-/) is not quite a good experience to use it in production, and we need to autoscale almost all of them.

I understand the concern from a dev POV, and it makes sense to start small for me too. I'll let you decide when to support more than one pool :)

Just curious, do you plan to do the next iteration soon? :)

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+	udvars := CloudInitTemplateData{
+		BootstrapTokenID:     os.Getenv("BOOTSTRAP_TOKEN_ID"),
+		BootstrapTokenSecret: os.Getenv("BOOTSTRAP_TOKEN_SECRET"),
+		APIServerEndpoint:    mgr.apiServerEndpoint,
```
(nitpicking, feel free to ignore)
Also, on a similar node, it seems weird to me that these are the selected variables for tempalte rendering, as these variables are needed for `kubeadm`. This is kind of hardcoding kubeadm, in some way.

This is not a big problem for our use case, but it seems like kubeadm details are leaking into a part that might be more generic. What are other providers doing, for example?

d-mo replied:
Bootstrap tokens are a standard way to bootstrap a new Kubernetes node and can be used without or beyond kubeadm. The above code makes them available to be used by the user-defined cloud-config template. I don't see any hardcoding nor anything leaking here.

Other providers leverage existing auto-scaling groups which are not available in Packet.

deitch replied:
Yes. Same thing with cluster-api implementations which often leverage the standard bootstrap tokens. If I am not mistaken, the latest kubeadm code actually uses the standard k8s tokens lib (i.e. from `github.com/kubernetes/kubernetes` to generate its own tokens.

rata replied:
Makes sense, thanks!

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+		APIServerEndpoint:    mgr.apiServerEndpoint,
+	}
+	ud := renderTemplate(string(cloudinit), udvars)
+	hn := "k8s-" + mgr.clusterName + "-" + nodegroup + "-" + randString8()
```
The hostname is quite hardcoded, we may want more flexibility to choose the servers names

invidian replied:
Also maybe we could use sequential numbers here, rather than random string suffic?

d-mo replied:
I added a new hostname-pattern option in cluster-autoscaler-cloud-config that makes it configurable. The random string suffix feels more standard in Kubernetes and in most cloud providers. Also, sequential numbers would reset every time cluster-autoscaler resets.

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+		OS:                    mgr.os,
+		ProjectID:             mgr.projectID,
+		BillingCycle:          mgr.billing,
+		UserData:              renderTemplate(ud, udvars),
```
Wasn't `ud` just rendered a few lines above?

d-mo replied:
Yes it was! Fixed.

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+	return nil
+}
+
+func createDevice(cr *DeviceCreateRequest) (*http.Response, error) {
```
Using more verbose logs doesn't increase verbosity significantly here (nor in the places calling this function), so it is difficult to debug.

For example, if you have a typo in the  facility name in the secret, the error shown doesn't say anything helpful and increasing verbosity doesn't show you the request (as it does happens with other HTTP request the autoscaler does, for example for the leader election) or anything like that might be more helpful to debug. It just panics() without a clear clue on why it is failing.

Would it be possible to have more verbose logs when verbosity is increased? It would have saved me quite some time, at least. I had a silly typo (using `scj1` instead of `sjc1` in the secret for facility name) and the code panic and increasing verbosity didn't help to find the error. I just had to look at everything hard until I found it.

d-mo replied:
I added some more log statements that print the HTTP requests and responses with when log level is 2 or higher. Let me know if you think there are more points in the code that could benefit from additional logging.

rata commented on cluster-autoscaler/cloudprovider/packet/examples/cluster-autoscaler-secret.yaml:
```diff
@@ -0,0 +1,50 @@
+
+  # Token ID and secret. Required.
+  token-id: 07401b
+  token-secret: f395accd246ae52d
```
It was not clear for me while reading the docs what were these value for. After reading the code and seeing the base64 encoded cloud-config is that I realized these are values for `kubeadm`. Maybe this can be better documented?

Also, the actual values won't work on a setup so maybe a placeholder might be better.

What do you think?

d-mo replied:
A better comment was indeed needed here. I updated it.

This yaml file is an example definition of secrets that should be customized on production. I tried to keep the example functional with as little user effort as possible. This particular secret adds the bootstrap token which is then available in the cloud-config template.

rata replied:
Yes, I understood it when reading the code. Now with the comment is easier to understand for all, thanks! :)

rata commented on cluster-autoscaler/cloudprovider/packet/examples/cluster-autoscaler-secret.yaml:
```diff
@@ -0,0 +1,50 @@
+  usage-bootstrap-signing: "true"
+
+  # Extra groups to authenticate the token as. Must start with "system:bootstrappers:"
+  auth-extra-groups: system:bootstrappers:kubeadm:default-node-token,system:bootstrappers:worker,system:bootstrappers:ingress
```
These fields were not clear for me, although I read the docs, what they meant.

d-mo replied:
The bootstrap token needs to get authenticated as a member of these groups and its usage needs to be defined. Since you're not using bootstrap tokens you can ignore this secret altogether.

rata replied:
Yes, I understood I could ignore it, thanks!

Sorry I wasn't clear. I think it might not be cleaer how to use this `auth-extra-groups` key from the README (was not mentioned) nor comments. A "docs" issue, let's say.

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+
+	udvars := CloudInitTemplateData{
+		BootstrapTokenID:     os.Getenv("BOOTSTRAP_TOKEN_ID"),
+		BootstrapTokenSecret: os.Getenv("BOOTSTRAP_TOKEN_SECRET"),
```
(nitpicking, feel free to ignore)
Seems weird to me to read those ENV variables unconditionally. That are only used if you want to use `kubeadm`. So, this feels a little hardcoded for kubeadm.

However, it doesn't fail if you don't use them, as `os.Getenv()` just returns an empty string if the var is not defined. So, there is no big problem from the user POV, just feels weird to me :)

rata reviewed (commented):
Hi @d-mo ! I'm Rodrigo, from Kinvolk. We had a quick chat the other day, don't know if you remember me :)

I played with the autoscaler code and could make it work, so great news! Thanks!

I left some comments in the code. Some are nit-picking, so I encourage you to ignore them if you don't agree (really!), and the most important one is about the support for only one nodepool limitation. I added some suggestion on how, maybe, that limitation can be removed.

Also, I notice there are not tests :-D

Let me know what you think

Thanks again,
Rodrigo

rata commented on cluster-autoscaler/cloudprovider/builder/builder_packet.go:
```diff
@@ -0,0 +1,42 @@
+	packet.ProviderName,
+}
+
+// DefaultCloudProvider for OpenStack-only build is OpenStack.
```
OpenStack? :)

d-mo replied:
Good catch, thanks! Fixed.

rata commented on cluster-autoscaler/cloudprovider/packet/examples/cluster-autoscaler-secret.yaml:
```diff
@@ -0,0 +1,50 @@
+  cloud-config: |-
+    [Global]
+    project-id=3d27fd13-0466-4878-be22-9a4b5595a3df
+    api-server-endpoint=147.75.102.15:6443
```
Maybe use a placeholder value here?

d-mo replied:
Indeed! Fixed

rata commented on cluster-autoscaler/cloudprovider/packet/examples/cluster-autoscaler-secret.yaml:
```diff
@@ -0,0 +1,50 @@
+stringData:
+  cloud-config: |-
+    [Global]
+    project-id=3d27fd13-0466-4878-be22-9a4b5595a3df
```
Maybe use a placeholder value here?

d-mo replied:
Fixed

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+	resp, err := createDevice(&cr)
+	if err != nil || resp.StatusCode > 299 {
+		// If reservation is preferred but not available, retry provisioning as on-demand
+		if reservation != "" && mgr.reservation == "prefer" {
```
This will just retry on any error if the reservation is configured as "prefer", right? Does't seem right to always retry, at least if the API returns a clear error for the "reservation not available" case (I think it does, judging for terraform errors, but haven't tried).

Or maybe I don't follow why any error would mean that the reservation is not available? Maybe I'm missing something?

d-mo replied:
Yes, that's right. If the response is not within the 200-299 range (which should mean in generic HTTP semantics that the create device request was successful) then cluster-autoscaler just panics and prints the error before dying, unless a reservation was set as preferred rather than required, in which case it will retry the request once, omitting the hardware_reservation_id param.

The Packet API returns a `503 Service Unavailable` response with the following body: `{"errors":["There are no available hardware reservations for selected plan and facility"]}`

Since we can't do much if the request fails anyway, I felt it was more future proof to assume that the reservation might have been the issue and retry without one, instead of looking for the exact error code and message which might conceivably change in the future. In any case, I don't have a very strong preference here. I can make it more tied to the current behavior of the Packet API if that's preferred.

rata replied:
Makes sense. If the error code is 503 when no reservation is done, it seems maybe better to retry ignoring it. It doesn't seem the best error code for this case :). Thanks!

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -0,0 +1,361 @@
+	}
+	defer resp.Body.Close()
+
+	ioutil.ReadAll(resp.Body)
```
Maybe a comment or assign to `_` so it is clear it is ignored?

d-mo replied:
Right, fixed. Also added some verbose logging of the response and some error handling.

Commit:
Modify nodal similarity rules.

exdx committed:
tweak: add CMD options to recommender entrypoint

feiskyer committed:
Fix race conditions when checking the VMSS size

k8s-ci-robot committed:
Merge pull request #2173 from feiskyer/fix-race

Fix race conditions when checking the VMSS size

k8s-ci-robot committed:
Merge pull request #2166 from Denton24646/recommender-dockerfile-add-CMD-args

tweak: add CMD options to recommender Dockerfile Entrypoint

exdx committed:
tweak: add CMD arg to Dockerfiles

bskiba committed:
Correct CMD for recommender

k8s-ci-robot committed:
Merge pull request #2176 from bskiba/dockerfile

Correct CMD for recommender

bskiba committed:
VPA deployments migrated to apps/v1

k8s-ci-robot committed:
Merge pull request #2177 from bskiba/dockerfile

VPA deployments migrated to apps/v1

bskiba committed:
e2e tests for v1 API

exdx committed:
fix: separate CMD args

k8s-ci-robot committed:
Merge pull request #2175 from Denton24646/tweak/update-Dockerfile-CMD-updater-admission-controller

tweak: add CMD to updater and admission-controller Dockerfile Entrypoint

k8s-ci-robot committed:
Merge pull request #2185 from bskiba/e2e-v1

e2e tests for v1 API

jackkleeman committed:
Use ioutil for certificate file reading.

Currently we read certificate files into a buffer of 5000 bytes. This
sets a max size for our certificates; above this max, we will get
invalid certs, most likely, leading to runtime bad certificate errors.

rata commented on cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go:
```diff
@@ -282,7 +301,8 @@ func createDevice(cr *DeviceCreateRequest) (*http.Response, error) {
 	url := "https://api.packet.net/projects/" + cr.ProjectID + "/devices"
 	jsonValue, _ := json.Marshal(cr)
-	klog.Infof("Creating new node")
+	klog.Infof("Creating new node yo")
```
`s/ yo//`? :)

d-mo replied:
Oops! Fixed

moriadry committed:
add test case for controller_fetcher and cluster_feeder
add test case for controller_fetcher
correct comments
typo

gtseres committed:
Fix typo

k8s-ci-robot committed:
Merge pull request #2188 from gtseres/master

Fix typo

bskiba committed:
Run VPA e2e tests for both apis even if one fails

k8s-ci-robot committed:
Merge pull request #2150 from cclauss/modernize-Python-2-codes

Use print() function in both Python 2 and Python 3

cclauss committed:
xrange() was removed in Python 3 in favor of range()

k8s-ci-robot committed:
Merge pull request #2190 from bskiba/e2e-v1

Run VPA e2e tests for both apis even if one fails

k8s-ci-robot committed:
Merge pull request #2187 from Moriadry/feature/coverage_input

add test case for controller_fetcher and cluster_feeder

k8s-ci-robot committed:
Merge pull request #2186 from jackkleeman/cert-file-read

Use ioutil for certificate file reading.

bskiba committed:
Add support for cronjobs in VPA

durch committed:
Fix typo in autoscaler azure example

- SubscriptionID placeholder does not match documentation or surrounding placeholders

k8s-ci-robot committed:
Merge pull request #2195 from bskiba/cronjob

Add support for cronjobs in VPA

Commit:
Modify the code as the simple solution proposed by MaciekPytel.

Jeffwan committed:
Update c5,i3en,m5,r5 instance types

k8s-ci-robot committed:
Merge pull request #2197 from Jeffwan/add_r5_ec2

Update c5,i3en,m5,r5 instance types

k8s-ci-robot committed:
Merge pull request #2191 from cclauss/fix-xrange

xrange() was removed in Python 3 in favor of range()

k8s-ci-robot committed:
Merge pull request #2196 from durch/patch-2

Fix typo in autoscaler azure example

Commit:
Refactor the corresponding code.

bskiba committed:
Update VPA dependencies to k8s 1.14.3

k8s-ci-robot committed:
Merge pull request #2149 from yan234280533/bugfix/nogroup_crash

nodeGroup judy IsNil to avoid crashed

k8s-ci-robot committed:
Merge pull request #2158 from kinvolk/rata/faq-use-metric-server-over-heapster

Use Metrics Server instead of Heapster in the FAQ

bskiba committed:
Switch VPA examples to use apps/v1 API for Deployment

k8s-ci-robot committed:
Merge pull request #2208 from bskiba/update-deps

Update VPA dependencies to k8s 1.14.3

k8s-ci-robot committed:
Merge pull request #2209 from bskiba/fix-example

Switch VPA examples to use apps/v1 API for Deployment

k8s-ci-robot committed:
Merge pull request #2171 from nilo19/t-qini-modify_nodal_similarity_rules

fix: ignore agentpool label when looking for similar node groups with Azure provider

Commit:
Add multinode config in readme.

k8s-ci-robot committed:
Merge pull request #2211 from nilo19/t-qini-add_multinode_config_readme

Add multinode config in readme.

Commit:
cluster-autoscaler: Replace ambiguous "delete node" wording in FAQ

"delete node" has a specific meaning in the context of the Kubernetes
API: deleting the node object. However, the cluster-autoscaler never
does that; it terminates the underlying instance and expects the
[cloud-node-controller](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#node-controller)
to remove the corresponding node from Kubernetes.

Replace all mentions of "deleting" a node with "terminating" to
disambiguate this.

Signed-off-by: Matthias Rampke <mr@soundcloud.com>

Commit:
cluster-autoscaler: add FAQ entry for the mechanics of "delete node"

This had me stumped for the better part of a day. While the
cluster-autoscaler "deletes" nodes, it does not actually delete the Node
object from the Kubernetes API.

In normal operations, with a well-configured cluster, this is a minor
point; however, when debugging why nodes do not get deleted, the
inconsistent terminology can be a major headache. This FAQ entry should
clarify the difference for anyone who needs to know.

Signed-off-by: Matthias Rampke <mr@soundcloud.com>

krzysztof-jastrzebski committed:
Fix bug in balancing processor. Cluster Autoscaler was stopping scaling
up when there was a multizonal pool with number of nodes exceeding limit for one zone.

kcmartin committed:
Add required selector to VPA deployment examples

k8s-ci-robot committed:
Merge pull request #2220 from kcmartin/vpa-add-selector-to-examples

Add required selector to VPA deployment examples for apps/v1

k8s-ci-robot committed:
Merge pull request #2219 from krzysztof-jastrzebski/fix_bug

Fix bug in balancing processor.

k8s-ci-robot committed:
Merge pull request #2217 from matthiasr/mr/does-not-delete-nodes

cluster-autoscaler FAQ: clarify what "deleting nodes" means in this context

jbartosik committed:
Fix v1beta1 VPA limit range e2e tests

I forgot to update them when fixing v1beta2 tests

k8s-ci-robot committed:
Merge pull request #2224 from jbartosik/bugfix-00

Fix v1beta1 VPA limit range e2e tests

jbartosik committed:
Remove unused import

k8s-ci-robot committed:
Merge pull request #2226 from jbartosik/bugfix-00

Remove unused import

vivekbagade committed:
Adding ScaleDownNodeProcessor

monaka committed:
Update README.md

Referring to `azure_manager.go` on 59b0087eeb, The default value of `vmType` is `vmTypeVMSS`.

Signed-off-by: Masaki Muranaka <monaka@monami-ya.com>

k8s-ci-robot committed:
Merge pull request #2234 from monaka/patch-1

Update README.md

jbartosik committed:
Don't attempt to delete aggregation from VPA if it doesn't exist

k8s-ci-robot committed:
Merge pull request #2238 from jbartosik/bugfix-00

Don't attempt to delete aggregation from VPA if it doesn't exist

tghartland committed:
Fix ldflags in CA Makefile

exdx committed:
fix: increase  duration regardless of backoff status

fix: update test

fix: run gofmt on test

fix: more intelligent duration increasing if in backoff already

set duration = backoffInfo.duration; update comments and test

set NodeGroup backoff duration = backoffInfo.duration

fatih committed:
cloudprovider: add godo, DigitalOcean API package

fatih committed:
cloudprovider: add DigitalOcean cloud provider

k8s-ci-robot committed:
Merge pull request #2245 from digitalocean/do-cloudprovider

cloudprovider: add DigitalOcean

Jeffwan committed:
Return error if aws fail to find nodegroup for node

towca committed:
Communicate the result of RemoveUnneededNodeGroups to ScaleDownStatusProcessor

towca committed:
Provide more info to ScaleUpStatusProcessor

Add info about considered and created nodegroups to
ScaleUpStatusProcessor

towca committed:
Provide ScaleUpStatusProcessor with info about all rejected node groups

Previously, it had info only about the ones that actually exist.

The changes to the eventing processor are done to keep its previous
behavior the same.

towca committed:
Fix the script used by CI to run unit tests after go mod changes

cluster-autoscaler switched from godeps to go mod, resulting in this
script being broken. This should fix issue 2251.

fatih committed:
cloudprovider/digitalocean: fix unit tests

The `godo/util` package is not used and wasn't changed during vendoring
of the godo package. Seems like Travis didn't picked up this in the
initial PR. Let's remove the `util/` folder as it's not needed and also
add a documentation to list the changes we need to make in the `godo`
package.

MaciekPytel committed:
Merge pull request #2256 from towca/jtuznik/unit-test-fix

Fix the script used by CI to run unit tests after go mod changes

k8s-ci-robot committed:
Merge pull request #2258 from digitalocean/fix-godo-tests

cloudprovider/digitalocean: fix unit tests

k8s-ci-robot committed:
Merge pull request #2254 from towca/jtuznik/remove-unneeded-nodes-result

Communicate the result of RemoveUnneededNodeGroups to ScaleDownStatus…

k8s-ci-robot committed:
Merge pull request #2255 from towca/jtuznik/create-node-group-result

Provide more info to ScaleUpStatusProcessor

bskiba committed:
Look at all versions of VPA objects.

This way we don't have to specify all the versions we are interested in.
Versioning is then controlled only at the crd level.

k8s-ci-robot committed:
Merge pull request #2261 from bskiba/gv

Look at all versions of VPA objects.

tghartland committed:
Add magnum to list of supported cloud providers

k8s-ci-robot committed:
Merge pull request #2247 from Jeffwan/fix-2161

Return error if aws fail to find nodegroup for node

realdimas committed:
VPA Recommender: Use WSS from Prometheus provider

Currently, VPA Recommender uses `container_memory_usage_bytes`
cAdvisor metric from Prometheus history provider.
This metric includes page cache into container memory usage total.
As a result, cache-heavy workloads keep getting larger and larger memory
recommendation.

This change takes memory held for page cache out of VPA's calculation
removing the primary driver behind memory over-requesting.

k8s-ci-robot committed:
Merge pull request #2233 from vivekbagade/surge

Adding ScaleDownNodeProcessor

k8s-ci-robot committed:
Merge pull request #2265 from tghartland/supported-provider-magnum

Add magnum to list of supported cloud providers

k8s-ci-robot committed:
Merge pull request #2244 from Denton24646/fix/update-duration-in-exponential-backoff

fix: increase exponential backoff duration regardless of backoff status

k8s-ci-robot committed:
Merge pull request #2259 from towca/jtuznik/rejected-node-groups-more-info

Provide ScaleUpStatusProcessor with info about all rejected node groups

k8s-ci-robot committed:
Merge pull request #2235 from piontec/fix/aws_spots_squashed

correctly handle lack of capacity of AWS spot ASGs

k8s-ci-robot committed:
Merge pull request #2266 from uswitch/use-wss-for-vpa-recommender-from-prometheus

VPA Recommender: Use WSS from Prometheus provider

k8s-ci-robot committed:
Merge pull request #2241 from tghartland/ldflags-2239

Fix linker flags in CA Makefile

bskiba committed:
Make VPA CRD specification OpenAPIV3 compliant

k8s-ci-robot committed:
Merge pull request #2267 from bskiba/open-api

Make VPA CRD specification OpenAPIV3 compliant

iaguis committed:
Handle adding several nodes at a time.

Log strings instead of byte arrays, so they're readable.

Fix number of log args
